### PR TITLE
Selection plan overlap

### PR DESCRIPTION
**Pull request by santipalenque**
https://tipit.avaza.com/project/view/304272#!tab=task-pane&groupby=ProjectSection&view=vertical&task=2860754&fileview=grid
https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskProject&view=vertical&task=2860698&fileview=grid
https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskProject&view=vertical&task=2860717&fileview=grid

### CODE REVIEW

File: src/actions/base-actions.js [hunk 1]
```diff
@@ -49,6 +50,26 @@ export const clearCurrentSummit = () => (dispatch, getState) => {
...
+
+    return getRequest(
+        null,
+        createAction(RECEIVE_SELECTION_PLAN),
+        `${window.API_BASE_URL}/api/v1/summits/${summitId}/selection-plans/${selectionPlanId}`,
+        selectionPlanErrorHandler
+    )(params)(dispatch);
+};
 
 export const getCurrentSelectionPlanPublic = (summit_id) => (dispatch, getState) => {
```
Comment (smarcet) [inline src/actions/base-actions.js:74]: this one should be removed if its not used anymore

File: src/reducers/base-reducer.js [hunk 2]
```diff
@@ -73,6 +73,8 @@ const baseReducer = (state = DEFAULT_STATE, action) => {
             return {...state, selectionPlan: null, summit: null, marketingSettings: null, submissionIsClosed: false, allSummitDocs: [], baseLoaded: false};
         case ERROR_RECEIVE_SUMMIT:
             return {...state, selectionPlan: null, summit: null, marketingSettings: null, submissionIsClosed: false};
+        case CLEAR_SELECTION_PLAN:
+            return {...state, selectionPlan: null, submissionIsClosed: false, baseLoaded: false};
         case RECEIVE_SELECTION_PLAN: {
             const entity = {...payload.response};
             return {...state, selectionPlan: entity, submissionIsClosed: false};
```
Comment (smarcet) [inline src/reducers/base-reducer.js:80]: here we do need to check is submission period is valid or not
selectionPlan.submission_begin_date <= now <= selectionPlan.submission_end_date

**Review by smarcet — CHANGES_REQUESTED** [review]
@santipalenque please review comments
regards

**Comment by smarcet** [comment]
@santipalenque also we need to setup the selection_plan_id on presentation save action
https://github.com/fntechgit/call-for-presentations/blob/b9e718c4c05f71a2bc35bad8dbf930b867a6ac11/src/reducers/presentation-reducer.js#L36

File: src/reducers/base-reducer.js (hunk 2 above)
Comment (smarcet) [reply src/reducers/base-reducer.js:80]: after discussion i will close this one

File: src/pages/plan-selection-page.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import {Dropdown} from 'openstack-uicore-foundation/lib/components'
+import history from "../history";
+import {connect} from "react-redux";
+import T from "i18n-react";
+import {clearSelectionPlan} from '../actions/base-actions';
+import '../styles/plan-selection-page.less';
+
+const PlanSelectionPage = ({summit, loading, match, clearSelectionPlan}) => {
+    const selPlans = summit?.selection_plans || [];
+    const availablePlans = selPlans.filter(sp => sp.is_enabled)
```
[the remark this answers is not in the document]
Comment (smarcet) [reply src/pages/plan-selection-page.js:24]: after discussion i will close this one

**Comment by smarcet** [comment]
@santipalenque also please review this one
https://tipit.avaza.com/project/view/304272#!tab=task-pane&groupby=ProjectSection&view=vertical&task=2864486&fileview=grid
regards

**Comment by smarcet** [comment]
@santipalenque one last change: 
i do think that here

https://github.com/fntechgit/call-for-presentations/blob/db4a595b89569b86642c79e1973e72c2cd78c08b/src/model/presentation.js#L45
Presentation.getStatus() method should be reviewed
those rules related to firstSelPlan/lastSelPlan does not applies anymore
we should look only at selection plan of the presentation itself
regards

**Comment by santipalenque** [comment]
> @santipalenque one last change: i do think that here
> 
> https://github.com/fntechgit/call-for-presentations/blob/db4a595b89569b86642c79e1973e72c2cd78c08b/src/model/presentation.js#L45
> 
> 
> Presentation.getStatus() method should be reviewed
> those rules related to firstSelPlan/lastSelPlan does not applies anymore
> we should look only at selection plan of the presentation itself
> regards

@smarcet can you provide the new logic for the presentation status ? Thanks!

File: src/layouts/primary-layout.js [hunk 4]
```diff
@@ -20,76 +20,52 @@ import ProfilePage from '../pages/profile-page'
 import SelectionProcessPage from '../pages/selection-process-page'
 import TracksGuidePage from '../pages/tracks-guide-page'
 import PresentationLayout from './presentation-layout'
-import { getAllFromSummit } from '../actions/base-actions';
+import {getSelectionPlan} from "../actions/base-actions";
 
-class PrimaryLayout extends React.Component {
+const PrimaryLayout = ({location, summit, speaker, member, match, selectionPlan, getSelectionPlan}) => {
+    const loggedUser = (speaker && speaker.id) ? speaker : member;
+    const selectionPlanIdParam = parseInt(match.params.selection_plan_id);
```
Comment (smarcet) [inline src/layouts/primary-layout.js:27]: could be useMemo here and then add it as dependency on useEffect?

File: src/layouts/primary-layout.js [hunk 5]
```diff
@@ -20,76 +20,52 @@ import ProfilePage from '../pages/profile-page'
...
-
-        if((!speaker || !speaker.id) && location.pathname !== `/app/${summit.slug}/profile` && !loading) {
-            return (
-                <Redirect exact to={{ pathname: `/app/${summit.slug}/profile` }}  />
-            );
+    useEffect(() => {
+        if (selectionPlan?.id !== selectionPlanIdParam) {
+            getSelectionPlan(summit.id, selectionPlanIdParam);
         }
+    }, [selectionPlan, match]);
```
Comment (smarcet) [inline src/layouts/primary-layout.js:43]: could we use more granular deps ?
like 
[selectionPlan?.Id, summit?.id, selectionPlanIdParam]

File: src/layouts/summit-layout.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import React, {useEffect} from 'react'
+import { connect } from 'react-redux';
+import { Switch, Route, Redirect } from 'react-router-dom';
+import {getAllFromSummit} from '../actions/base-actions';
+import PrimaryLayout from "./primary-layout";
+import PlanSelectionPage from "../pages/plan-selection-page";
+
+const SummitLayout = ({summit, loading, match, speaker, location, getAllFromSummit}) => {
+    useEffect(() => {
+        let summitSlug = match.params.summit_slug;
```
Comment (smarcet) [inline src/layouts/summit-layout.js:23]: could we redeclared this outside as useMemo ?
const summitSlug = useMemo(() =>  match.params.summit_slug, [ match?.params?.summit_slug])

File: src/layouts/summit-layout.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        if (summitSlug !== summit?.slug) {
+            getAllFromSummit(summitSlug);
+        }
+    }, [summit, match]);
```
Comment (smarcet) [inline src/layouts/summit-layout.js:27]: coiuld we add more granular deps here?
like
[summitSlug, summit?.slug]

File: src/layouts/summit-layout.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import { connect } from 'react-redux';
+import { Switch, Route, Redirect } from 'react-router-dom';
+import {getAllFromSummit} from '../actions/base-actions';
+import PrimaryLayout from "./primary-layout";
+import PlanSelectionPage from "../pages/plan-selection-page";
+
+const SummitLayout = ({summit, loading, match, speaker, location, getAllFromSummit}) => {
+    useEffect(() => {
+        let summitSlug = match.params.summit_slug;
+        if (summitSlug !== summit?.slug) {
```
Comment (smarcet) [inline src/layouts/summit-layout.js:24]: i do think that here we need to call first to  this.props.clearSelectionPlan(); to clear the current plan state ?

File: src/layouts/primary-layout.js (hunk 4 above)
Comment (santipalenque) [reply src/layouts/primary-layout.js:27]: useMemo is for expensive calculations, this is just a parse, I don't think we need to memoize

File: src/layouts/primary-layout.js (hunk 5 above)
Comment (santipalenque) [reply src/layouts/primary-layout.js:43]: we don't need summit.id here , I'll add the rest

File: src/layouts/summit-layout.js (hunk 8 above)
Comment (santipalenque) [reply src/layouts/summit-layout.js:24]: selection plan is cleared in the reducer when we pull a new summit

**Review by smarcet — APPROVED** [review]
looks good to me

**Pull request merged**